Fix scaffold generator: handle model name when plural form is unknown (aka the "Pokemon" bug)

callingmedic911 opened the pull request:
Closes #684.

This attempts to fix the crash of generators if the model name passed is an uncountable noun (plural === singular). The approach is to let the user decide different variants (like @cannikin suggested in https://github.com/redwoodjs/redwood/issues/684#issuecomment-862563971).

### To do:
- Prompt:
    - [x] Ask plural variant from user
- Generators to update:
   - [x] scaffold
   - [x] sdl
   - [x] service
- [x] Update test case
- [ ] Update docs

callingmedic911 commented:
@cannikin @thedavidprice 
- If we ask the user both singular and plural variant, it could be completely different from what was initially provided to CLI. That means we would need to maintain the model name separately (to `getSchema()`) and scaffold name separately to generate template files. Doable (still need to check changes/impact) but it looks confusing from experience standpoint. This happens only for uncountable ones anyways, so for now, I've assumed the provided one is singular and asking the user to plural variant only. Thoughts?

    Example: 
```
yarn rw g scaffold Pokemon
-> Enter singular: Something
-> Enter plural: Else

getSchema will fail to find "Something" model in the schema.
```
- Not able to use `prompt` inside `Listr`, at least not properly. Here's what's happening: https://s.tape.sh/sg0sLUnd
Currently added check before creating `tasks`. Any idea on this one? 

Glad to change my decisions to whatever we decide is best.

cannikin commented:
You can see what I did here: https://github.com/redwoodjs/redwood/blob/main/packages/cli/src/commands/setup/auth/auth.js#L246

I prompt first before starting Listr, so as long as it's the first step it should be doable!

callingmedic911 commented:
> I prompt first before starting Listr, so as long as it's the first step it should be doable!

Right. Doing exactly that. :)

thedavidprice commented:
> This happens only for uncountable ones anyways, so for now, I've assumed the provided one is singular and asking the user to plural variant only. Thoughts?

^^ I 💯 agree with this approach. Since singular === model, the user could easily quit and start the process again after using a different model name. 

This PR should focus on the plural only.

Lastly, I think it's appropriate to suggest the plural, which user could accept or update. e.g.:
```
yarn rw g scaffold Pokemon
-> We don't know the plural form of Pokemon.
-> Enter plural (or accept suggestion): Pokemons 
```

Might not be necessary. So up to you.

callingmedic911 commented:
> Lastly, I think it's appropriate to suggest the plural, which user could accept or update. e.g.:

Yes! Here's a demo: https://s.tape.sh/kxxjZXIQ
I've marked the PR, please do look at the phrases for message/errors I've used. Let me know if I should change something.

I'll start on the doc as soon we finalize these.

thedavidprice reviewed (changes requested):
@callingmedic911 this is looking great! The functionality seems right on.

Here's a recording of my run-through to view:
https://s.tape.sh/Y3wWefPK

It's an edge-case, but I suggest adding a check for `empty`.

> We don't know the plural form of "pokemon" which is different from its singular.

^^Maybe even simpler --> `Cannot determine the plural of "pokemon". To continue, the generator requires a unique plural form:`

Lastly, do you think there could be a reason to add a command option to automatically accept the plural suggestion? I don't believe it's necessary, but it seems worth discussing. (Thoughts about this @dthyresson?)

callingmedic911 commented:
> It's an edge-case, but I suggest adding a check for `empty`.

Great catch! I thought it should fairly straightforward to check for empty string but turns out it is not an empty string if we clear the default value using option+backspace. It's returning [control-char ETB](https://en.wikipedia.org/wiki/End-of-Transmission-Block_character). This should be handled at library level - from `prompt` - but for now, I've added a check in the validator along with other space characters, and removed this char if encountered in the plural name (option + backspace and start typing, it'll output \u0017whatyoutyped). It feels a little hackish, I open to any suggestion. :)

This was very tricky to debug because the console was printing blank. Very interesting stuff I learned today. 😁

> ^^Maybe even simpler --> `Cannot determine the plural of "pokemon". To continue, the generator requires a unique plural form:`

Done.

> Lastly, do you think there could be a reason to add a command option to automatically accept the plural suggestion? I don't believe it's necessary, but it seems worth discussing.

From https://github.com/redwoodjs/redwood/issues/684#issuecomment-862636594, I think if for some reason I want to put this command in any scripts, it'll be a little difficult to provide input.
> if I want to throw-in a couple of rw command in a shell script, it would a bit more challenging to manage, probably will have to use something like expect.

thedavidprice commented on packages/cli/src/commands/generate/helpers.js:
```diff
@@ -218,12 +224,14 @@ export const ensureUniquePlural = async (model) => {
     type: 'text',
     name: 'plural',
-    message: `We don't know the plural form of "${model}" which is different from its singular.\n=> Enter plural (or accept suggestion):`,
+    message: `Cannot determine the plural of "${model}" "pokemon". To continue, the generator requires a unique plural form:`,
```
```suggestion
    message: `Cannot determine the plural of "${model}". To continue, the generator requires a unique plural form:`,
```

callingmedic911 replied:
Oops. 🙈

thedavidprice commented on packages/cli/src/commands/generate/helpers.js:
```diff
@@ -192,3 +197,45 @@ export const forcePluralizeWord = (word) => {
+  const promptResult = await prompts({
+    type: 'text',
+    name: 'plural',
+    message: `Cannot determine the plural of "${model}". \nTo continue, the generator requires a unique plural form:`,
```
@callingmedic911 This is looking great and is so very close! Just one final thing I noticed -- when running the `destroy` command, the message above is out of context and no longer makes complete sense. (see screenshot below) I suggest that when used for `destroy`, something like the following message should display instead:
```js
message: `Cannot determine the plural of "${model}" originally used to generate scaffolding. \nTo continue, the destroy command requires the plural form:`,
```

<img width="612" alt="Screen Shot 2021-06-21 at 10 16 43 AM" src="https://user-images.githubusercontent.com/2951/122802856-bdf93100-d27a-11eb-81ee-32c8ec5973c7.png">

callingmedic911 replied:
Oh right! The part in the message "originally used to generate scaffolding" feels very specific to `scaffold`. Do you think it would be okay to display this for `destory sdl` or `service` too?

Thanks a lot for the help @thedavidprice! 🙂

callingmedic911 replied:
Updated the message for destroyer prompt like you suggested. But do let me know your thoughts on above. :)

thedavidprice reviewed (changes requested):
just one minor change. see inline comment

thedavidprice commented:
@cannikin this one is gtg for me with the exception of one messaging case for the `destroy` command. Do you want to take a look and add a review before merging?

Here's a recording of the current behavior:
https://s.tape.sh/5agAGUuF

thedavidprice commented:
Huge thank you @callingmedic911 🚀 This bug was outstanding for far too long.

There are no Doc updates necessary, so everything is now complete!

mojombo commented:
@callingmedic911, I was on a live video coding session when I was first bit by the "Pokemon" bug so I am very excited to have this fixed to save myself future embarrassment. =) Amazing work, thank you!!!
